### PR TITLE
added first solution

### PR DESCRIPTION
**Pull request by IhorRome**

### CODE REVIEW

File: src/main/java/practice/StreamPractice.java [hunk 1]
```diff
@@ -23,7 +36,11 @@ public int findMinEvenNumber(List<String> numbers) {
      * But before that subtract 1 from each element on an odd position (having the odd index).
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        return IntStream.range(0, numbers.size())
+                .mapToObj(i -> (i % 2 != 0) ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(integer -> integer % 2 != 0)
+                .mapToDouble(Integer::doubleValue)
+                .average().getAsDouble();
```
Comment (LayRipper) [inline src/main/java/practice/StreamPractice.java:43]:
```suggestion
                .mapToDouble(i -> (i % 2 != 0) ? numbers.get(i) - 1 : numbers.get(i))
                .filter(integer -> integer % 2 != 0)
                .average().getAsDouble();
```

File: src/main/java/practice/StreamPractice.java [hunk 2]
```diff
@@ -50,7 +71,11 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      */
     public List<People> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<People> peopleList) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= fromAge)
+                .filter(person -> person.getSex() == People.Sex.MAN ? person.getAge() <= maleToAge
+                        : person.getAge() <= femaleToAge)
```
Comment (LayRipper) [inline src/main/java/practice/StreamPractice.java:77]: These two filters can be merged into one as well

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -14,7 +18,16 @@
...
-        return 0;
+        return numbers.stream()
+                .map(str -> str.split(","))
+                .flatMap(Arrays::stream)
+                .map(Integer::parseInt)
+                .mapToInt(Integer::intValue)
+                .filter(i -> i % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list: "
+                        + numbers.toString()));
```
Comment (LayRipper) [inline src/main/java/practice/StreamPractice.java:29]: Unnecessary toString call 
```suggestion
                .orElseThrow(() -> new RuntimeException("Can't get min value from list: "
                        + numbers.));
```

**Review by LayRipper — CHANGES_REQUESTED** [review]

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -59,7 +83,12 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      * return the names of all cats whose owners are women from `femaleAge` years old inclusively.
      */
     public List<String> getCatsNames(List<People> peopleList, int femaleAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex() == People.Sex.WOMEN && p.getAge() > femaleAge)
+                .map(People::getCats)
```
Comment (LayRipper) [inline src/main/java/practice/StreamPractice.java:88]: * return the names of all cats whose owners are women from `femaleAge` years old inclusively.
```suggestion
                .filter(p -> p.getSex() == People.Sex.WOMEN && p.getAge() >= femaleAge)
                .map(People::getCats)
```

**Review by LayRipper — CHANGES_REQUESTED** [review]

**Review by LayRipper — APPROVED** [review]

File: src/main/java/practice/CandidateValidator.java [hunk 5]
```diff
@@ -1,5 +1,22 @@
 package practice;
 
-public class CandidateValidator {
+import java.util.function.Predicate;
+import model.Candidate;
 
+public class CandidateValidator implements Predicate<Candidate> {
+
```
Comment (Resci) [inline src/main/java/practice/CandidateValidator.java:7]:
```suggestion
```

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -14,7 +18,16 @@
...
+        return numbers.stream()
+                .map(str -> str.split(","))
+                .flatMap(Arrays::stream)
+                .map(Integer::parseInt)
+                .mapToInt(Integer::intValue)
+                .filter(i -> i % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list: "
+                        + numbers));
+
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:30]:
```suggestion
```

File: src/main/java/practice/StreamPractice.java [hunk 7]
```diff
@@ -14,7 +18,16 @@
      * "Can't get min value from list: < Here is our input 'numbers' >"
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .map(str -> str.split(","))
+                .flatMap(Arrays::stream)
+                .map(Integer::parseInt)
+                .mapToInt(Integer::intValue)
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:25]: Two map operations is enough here

File: src/main/java/practice/StreamPractice.java [hunk 8]
```diff
@@ -23,7 +36,10 @@ public int findMinEvenNumber(List<String> numbers) {
      * But before that subtract 1 from each element on an odd position (having the odd index).
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        return IntStream.range(0, numbers.size())
+                .mapToDouble(i -> (i % 2 != 0) ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(integer -> integer % 2 != 0)
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:41]: Comparing two double variables through == or != you may have bugs. Let's think about how we can avoid them

File: src/main/java/practice/StreamPractice.java [hunk 9]
```diff
@@ -23,7 +36,10 @@ public int findMinEvenNumber(List<String> numbers) {
...
+                .average().getAsDouble();
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:42]:
```suggestion
                .average()
                .getAsDouble();
```

File: src/main/java/practice/StreamPractice.java [hunk 10]
```diff
@@ -59,7 +83,12 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      * return the names of all cats whose owners are women from `femaleAge` years old inclusively.
      */
     public List<String> getCatsNames(List<People> peopleList, int femaleAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex() == People.Sex.WOMEN && p.getAge() >= femaleAge)
+                .map(People::getCats)
+                .flatMap(Collection::stream)
+                .map(Cat::getName)
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:90]: Same, two map operations is enough here

**Review by Resci — CHANGES_REQUESTED** [review]

File: src/main/java/practice/StreamPractice.java [hunk 11]
```diff
@@ -1,80 +1,63 @@
...
-     * for example, `Arrays.asList( new People(«Victor», 16, Sex.MAN),
-     * new People(«Helen», 42, Sex.WOMEN))`,
-     * select from the List only men whose age is from `fromAge` to `toAge` inclusively.
-     * <p>
-     * Example: select men who can be recruited to army (from 18 to 27 years old inclusively).
-     */
     public List<People> selectMenByAge(List<People> peopleList, int fromAge, int toAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= fromAge
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:33]: `If you have a long boolean expression you may create your own implementation of Predicate and use it in your filter.` - checklist

File: src/main/java/practice/StreamPractice.java [hunk 12]
```diff
@@ -1,80 +1,63 @@
...
-     * If there is no needed data throw RuntimeException with message
-     * "Can't get min value from list: < Here is our input 'numbers' >"
-     */
+
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .flatMap(s -> Arrays.stream(s.split(",")))
+                .mapToInt(Integer::parseInt)
+                .filter(i -> i % 2 == 0)
```
Comment (Resci) [inline src/main/java/practice/StreamPractice.java:17]: `n`, `i`, `p` - is not informative enough
let's better use `number`, `index`, `person`

**Review by Resci — CHANGES_REQUESTED** [review]

**Review by Resci — CHANGES_REQUESTED** [review]
test

**Review by Resci — CHANGES_REQUESTED** [review]
test

**Review by Resci — APPROVED** [review]

**Pull request closed without merging**